### PR TITLE
Open Registration for torch::deploy Builtins

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66021
* __->__ #65953

A torch::deploy builtin is a python library which is loaded directly into the torch::deploy embedded interpreters. Due to a bug regarding having duplicate symbols, we need to explicitly load python binaries as builtins (ie. numpy, pandas, etc.) as buitins to torch::deploy.

Previously if people want to add a torch::deploy builtin, they need to change torch::deploy internal code (interpreter_impl.cpp) to register the python part as frozen modules and C++ part as builtin modules. This is not convenient and error prone. We want to add open registration support for torch::deploy builtins so that people only need to add one effective line of code in there *library code* to complete the registration.

Here is an example to registry numpy as torch::deploy builtins:
  REGISTER_TORCH_DEPLOY_BUILTIN(numpy, numpy_frozen_modules, <list of name, PyInit function pairs>)

This diff supports open registration of frozen modules. It's the first step to achieve the plan above.

Differential Revision: [D31321562](https://our.internmc.facebook.com/intern/diff/D31321562/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31321562/)!

### CODE REVIEW

**Comment by shunting314** [comment]
I checked the CI failure: linux-xenial-py3.6-gcc5.4 / test (backwards_compat, 1, 1, linux.2xlarge) 
It complains about backward compatibility issue specifically for softmax and log_softmax ops. I don't think it's related to this PR.

**Review by suo — CHANGES_REQUESTED** [review]
Excellent work! I only have minor comments. Overall nit about style: we should try to follow consistent (read: FB) style. So, camel case instead of snake for most things.

File: torch/csrc/deploy/interpreter/interpreter_impl.cpp [hunk 1]
```diff
@@ -191,69 +191,43 @@ warnings.simplefilter("ignore")
...
+
+  auto* frozenpython = builtin_registry::get_item("frozenpython");
+  auto* frozentorch = builtin_registry::get_item("frozentorch");
+  // Check frozenpython+frozentorch together since in OSS frozenpython is empty
+  // and frozentorch contains stdlib+torch, while in fbcode they are separated
+  // due to thirdparty2 frozenpython. No fixed number of torch modules to check
+  // for, but there should be at least one.
   TORCH_INTERNAL_ASSERT(
-      a + b > NUM_FROZEN_PY_STDLIB_MODULES + 1,
+      frozenpython != nullptr && frozentorch != nullptr &&
```
Comment (suo) [inline torch/csrc/deploy/interpreter/interpreter_impl.cpp:220]: This is just a sanity check; I wonder if we should avoid exposing `num_modules` and `get_item` and instead handle this check (and the registration of frozenpython/frozentorch) inside the builtin registry instead of doing it here.

**Comment by suo** [comment]
Oh, also I suggest rebasing after master; https://github.com/pytorch/pytorch/pull/65861 probably threw up some conflicts

**Comment by wconstab** [comment]
wow, excited to see this!

File: torch/csrc/deploy/builtin_registry.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      name,
+      num_modules);
+}
+
+builtin_registry* builtin_registry::get() {
+  static builtin_registry _registry;
+  return &_registry;
+}
+
+void builtin_registry::register_builtin(
```
[the remark this answers is not in the document]
Comment (shunting314) [reply torch/csrc/deploy/builtin_registry.cpp:30]: yes, it's safer to add mutex to guard the accesses. But giving it another think, I feel maybe we don't need that?

Typical use of the BuilitinRegistry is we call macro REGISTER_TORCH_DEPLOY_BUILTIN in the global scope of a source file. The macro just create an anonymous registerer object. The constructor of those registerer object will call builtin_registry::register_builtin. Those calls to the registerer's constructor will be collected to .init section of the embedded interpreter and being executed when we call dlopen.

So I think for this typical scenario, things are done in a single thread. Adding mutex for all accesses will make the code a bit complex.

Let me know if miss something

[the remark this answers is not in the document]
Comment (suo) [reply torch/csrc/deploy/builtin_registry.cpp:30]: this is true if we assume that the macro will only ever be called during static initialization, but it's not necessarily the case. Our experience with custom operators (which are registered in a similar way) suggests that sometimes people call them during runtime as well. Since we don't control it, better to be safe I think?

[the remark this answers is not in the document]
Comment (shunting314) [reply torch/csrc/deploy/builtin_registry.cpp:30]: > or hmmm, I guess since they have to be part of the embedded interpreter build it doesn't make sense for them to be called outside of static initialization? Is that what you're saying?

yep. Also there is very little chance that people can register a torch::deploy builtin at runtime. Since if they really want to do that, they need to do that before we iterate thru the list of registered builtins during initialization in interpreter_impl.cpp.

**Review by suo — APPROVED** [review]
Awesome stuff!

**Review by suo — APPROVED** [review]
